File: classNotes/1 Java Core Examples/4_Operators.java
/*
 * Topic: Operators
 * Core idea: the usual operator families, plus bitwise operators given their
 * own section - they're tested heavily in placement rounds and easy to skip
 * in an intro course.
 */

import java.util.ArrayList;
import java.util.List;

public class Operators {

    public static void main(String[] args) {

        int a = 20, b = 6;

        // Arithmetic
        System.out.println(a + b);
        System.out.println(a - b);
        System.out.println(a * b);
        System.out.println((double) a / b);   // true division needs an explicit cast
        System.out.println(a / b);             // integer division truncates
        System.out.println(a % b);
        System.out.println(Math.pow(a, b));    // no ** operator in Java

        // Comparison
        System.out.println(a == b);
        System.out.println(a > b);

        // Logical
        boolean x = true, y = false;
        System.out.println(x && y);
        System.out.println(x || y);
        System.out.println(!x);

        // Bitwise - frequently tested, rarely covered in an intro course
        System.out.println(a & b);
        System.out.println(a | b);
        System.out.println(a ^ b);
        System.out.println(~a);
        System.out.println(a << 1);   // equivalent to a * 2
        System.out.println(a >> 1);   // equivalent to a / 2

        // Assignment
        int num = 10;
        num += 5;
        num *= 2;
        System.out.println(num);

        // Membership - no `in` keyword in Java
        System.out.println("Python".contains("P"));

        // Identity vs equality
        List<Integer> list1 = new ArrayList<>(List.of(1, 2, 3));
        List<Integer> list2 = list1;
        List<Integer> list3 = new ArrayList<>(List.of(1, 2, 3));
        System.out.println(list1 == list2);        // same object
        System.out.println(list1 == list3);         // different objects
        System.out.println(list1.equals(list3));    // same content
    }
}

File: classNotes/2 Python Core Examples/1 fundamentals/4 Operators.py
"""
Topic: Operators
Author: Digital Flash Notes
"""

a = 20
b = 6


# Arithmetic Operators

print(a + b)
print(a - b)
print(a * b)
print(a / b)
print(a // b)
print(a % b)
print(a ** b)


# Comparison Operators

print(a == b)
print(a != b)
print(a > b)
print(a < b)
print(a >= b)
print(a <= b)


# Logical Operators

x = True
y = False

print(x and y)
print(x or y)
print(not x)


# Assignment Operators

num = 10

num += 5
print(num)

num -= 2
print(num)

num *= 3
print(num)


# Membership Operators

language = "Python"

print("P" in language)
print("Java" in language)


# Identity Operators

list1 = [1, 2, 3]
list2 = list1
list3 = [1, 2, 3]

print(list1 is list2)
print(list1 is list3)

print(list1 == list3)